### PR TITLE
cost calculation - workers running as master-subtasks

### PR DESCRIPTION
**Pull request by mkurek**
* changed calculation costs worker jobs to master job and subtasks schema - master job run subtasks to calculate costs for each day in period of time (each subtask is responsible for single day). Then master job collects result from subtasks, process them (ex. create DailyCost instances) and at the end, when all subtasks are done, save calculated results to the database.
* from now Scrooge requires (at least) two workers running on 'scrooge_costs' queue (or default).

### CODE REVIEW

**Comment by zefciu** [comment]
:ok:

File: src/ralph_scrooge/plugins/cost/collector.py [hunk 1]
```diff
@@ -157,17 +160,36 @@ def process(
...
+        """
+        # create daily costs instances
         daily_costs = self._create_daily_costs(date, costs, forecast)
-        with transaction.commit_on_success():
-            self._delete_daily_costs(date, forecast)
-            self._save_costs(date, daily_costs, forecast)
+        self._delete_daily_costs(date, forecast)
+        self._save_costs(daily_costs)
+        self._update_status(date, forecast)
         logger.info('Costs saved for date {}'.format(date))
```
Comment (kula1922) [inline src/ralph_scrooge/plugins/cost/collector.py]: as we discussed

File: src/ralph_scrooge/utils/worker_job.py [hunk 2]
```diff
@@ -78,10 +80,17 @@ def run_on_worker(self, **kwargs):
                 (progress, job.id, data),
                 timeout=self.cache_timeout,
             )
-        return progress, data
+        if self._return_job_meta:
+            return progress, data, job.meta
+        else:
```
Comment (kula1922) [inline src/ralph_scrooge/utils/worker_job.py]: needless 'else' statement

File: src/ralph_scrooge/plugins/cost/collector.py [hunk 3]
```diff
@@ -157,17 +160,36 @@ def process(
...
             plugins,
         )
+        logger.info('Costs calculated for date {}'.format(date))
+        return costs
+
+    def save_costs(self, date, forecast, costs):
+        """
+        Save costs for single date.
+
+        :param costs: costs processed by collector and and plugins
```
Comment (kula1922) [inline src/ralph_scrooge/plugins/cost/collector.py]: double 'and'

File: src/ralph_scrooge/views/monthly_costs.py [hunk 4]
```diff
@@ -161,15 +114,139 @@ def accept_costs(self, start, end, forecast):
...
+        :type end: datetime.date
+        """
+        for day in rrule.rrule(rrule.DAILY, dtstart=start, until=end):
+            self._clear_cache(day=day, **kwargs)
+        self._clear_cache(start=start, end=end, **kwargs)
+
+    @classmethod
+    def _check_subjobs(cls, statuses, start, end, **kwargs):
+        """
+        Check subjobs (jobs for single day) statuses.
```
Comment (kula1922) [inline src/ralph_scrooge/views/monthly_costs.py]: what is statuses? plx descripe params

File: src/ralph_scrooge/views/monthly_costs.py [hunk 5]
```diff
@@ -161,15 +114,139 @@ def accept_costs(self, start, end, forecast):
...
+        Save costs between start and end.
+        """
+        collector = Collector()
+        collector.save_period_costs(start, end, forecast, data)
+
+    @classmethod
+    def _process_daily_result(self, data, date, forecast):
+        """
+        Process results from subtask jobs (single day) - create daily costs
+        instances.
```
Comment (kula1922) [inline src/ralph_scrooge/views/monthly_costs.py]: what is data? dito

File: src/ralph_scrooge/views/monthly_costs.py (hunk 4 above)
Comment (mkurek) [inline src/ralph_scrooge/views/monthly_costs.py]: done

File: src/ralph_scrooge/views/monthly_costs.py (hunk 5 above)
Comment (mkurek) [inline src/ralph_scrooge/views/monthly_costs.py]: done

**Comment by kula1922** [comment]
:ok:

**Pull request merged**